File: flin_shopify_analytics_mcp/mcp_server.py
"""Minimal JSON-RPC/MCP server implementation."""

from __future__ import annotations

from typing import Any

from .config import Config, load_config
from .shopify_client import ShopifyClient
from .tools import TOOL_DEFINITIONS, call_tool


JSONRPC_VERSION = "2.0"
DEFAULT_PROTOCOL_VERSION = "2025-03-26"
SERVER_INFO = {"name": "flin-shopify-analytics-mcp", "version": "0.2.1"}


class ShopifyAnalyticsMcpServer:
    def __init__(self, client: ShopifyClient | None = None, config: Config | None = None):
        if client is not None:
            self.client = client
        else:
            resolved_config = config or load_config()
            self.client = ShopifyClient(resolved_config)
        self.initialized = False

    @staticmethod
    def _is_notification(message: dict[str, Any]) -> bool:
        return "id" not in message

    @staticmethod
    def _error(code: int, message: str, data: Any | None = None) -> dict[str, Any]:
        payload = {"code": code, "message": message}
        if data is not None:
            payload["data"] = data
        return payload

    def handle_message(self, message: dict[str, Any] | None) -> dict[str, Any] | None:
        if not isinstance(message, dict):
            return {"jsonrpc": JSONRPC_VERSION, "id": None, "error": self._error(-32600, "Invalid Request")}

        method = message.get("method")
        if not isinstance(method, str):
            if self._is_notification(message):
                return None
            return {
                "jsonrpc": JSONRPC_VERSION,
                "id": message.get("id"),
                "error": self._error(-32600, "Invalid Request"),
            }

        try:
            result = self._dispatch(method, message.get("params") or {})
            if self._is_notification(message):
                return None
            return {"jsonrpc": JSONRPC_VERSION, "id": message["id"], "result": result}
        except Exception as exc:  # pragma: no cover
            if self._is_notification(message):
                return None
            code = getattr(exc, "code", -32000)
            text = str(exc)
            return {"jsonrpc": JSONRPC_VERSION, "id": message.get("id"), "error": self._error(code, text)}

    def _assert_initialized(self, method: str) -> None:
        if not self.initialized:
            error = RuntimeError(f"Server not initialized. Call initialize before {method}.")
            setattr(error, "code", -32002)
            raise error

    def _dispatch(self, method: str, params: dict[str, Any]) -> dict[str, Any]:
        if method == "initialize":
            self.initialized = True
            return {
                "protocolVersion": params.get("protocolVersion") or DEFAULT_PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "serverInfo": SERVER_INFO,
            }
        if method == "notifications/initialized":
            return {}
        if method == "ping":
            return {}
        if method == "tools/list":
            self._assert_initialized(method)
            return {"tools": TOOL_DEFINITIONS}
        if method == "tools/call":
            self._assert_initialized(method)
            return self._handle_tool_call(params)
        if method == "resources/list":
            self._assert_initialized(method)
            return {"resources": []}
        if method == "prompts/list":
            self._assert_initialized(method)
            return {"prompts": []}

        error = RuntimeError(f"Method not found: {method}")
        setattr(error, "code", -32601)
        raise error

    def _handle_tool_call(self, params: dict[str, Any]) -> dict[str, Any]:
        name = params.get("name")
        if not isinstance(name, str) or not name:
            error = RuntimeError("Invalid params: tools/call requires a string `name`.")
            setattr(error, "code", -32602)
            raise error

        try:
            return call_tool(self.client, name, params.get("arguments") or {})
        except Exception as exc:
            return {"content": [{"type": "text", "text": f"Tool failed: {exc}"}], "isError": True}

File: src/mcp-server.js
import { loadConfig } from "./config.js";
import { ShopifyClient } from "./shopify-client.js";
import { TOOL_DEFINITIONS, callTool } from "./tools.js";

const JSONRPC_VERSION = "2.0";
const DEFAULT_PROTOCOL_VERSION = "2025-03-26";
const SERVER_INFO = {
  name: "flin-shopify-analytics-mcp",
  version: "0.1.0"
};

function isNotification(message) {
  return message.id === undefined;
}

function makeError(code, message, data) {
  const error = { code, message };
  if (data !== undefined) {
    error.data = data;
  }
  return error;
}

export class ShopifyAnalyticsMcpServer {
  constructor({ client, config, fetchImpl } = {}) {
    this.client = client;
    if (!this.client) {
      const resolvedConfig = config || loadConfig();
      this.client = new ShopifyClient(resolvedConfig, { fetchImpl });
    }
    this.initialized = false;
  }

  async handleMessage(message) {
    if (!message || typeof message !== "object") {
      return {
        jsonrpc: JSONRPC_VERSION,
        id: null,
        error: makeError(-32600, "Invalid Request")
      };
    }

    if (typeof message.method !== "string") {
      if (isNotification(message)) {
        return null;
      }
      return {
        jsonrpc: JSONRPC_VERSION,
        id: message.id ?? null,
        error: makeError(-32600, "Invalid Request")
      };
    }

    try {
      const result = await this.dispatch(message.method, message.params ?? {});
      if (isNotification(message)) {
        return null;
      }
      return {
        jsonrpc: JSONRPC_VERSION,
        id: message.id,
        result
      };
    } catch (error) {
      if (isNotification(message)) {
        return null;
      }
      const code = Number.isInteger(error.code) ? error.code : -32000;
      return {
        jsonrpc: JSONRPC_VERSION,
        id: message.id ?? null,
        error: makeError(code, error.message || "Internal error")
      };
    }
  }

  async dispatch(method, params) {
    switch (method) {
      case "initialize":
        this.initialized = true;
        return {
          protocolVersion: params.protocolVersion || DEFAULT_PROTOCOL_VERSION,
          capabilities: {
            tools: {}
          },
          serverInfo: SERVER_INFO
        };
      case "notifications/initialized":
        return {};
      case "ping":
        return {};
      case "tools/list":
        this.assertInitialized(method);
        return { tools: TOOL_DEFINITIONS };
      case "tools/call":
        this.assertInitialized(method);
        return this.handleToolCall(params);
      case "resources/list":
        this.assertInitialized(method);
        return { resources: [] };
      case "prompts/list":
        this.assertInitialized(method);
        return { prompts: [] };
      default:
        throw { code: -32601, message: `Method not found: ${method}` };
    }
  }

  assertInitialized(method) {
    if (!this.initialized) {
      throw { code: -32002, message: `Server not initialized. Call initialize before ${method}.` };
    }
  }

  async handleToolCall(params) {
    const name = params?.name;
    if (!name || typeof name !== "string") {
      throw { code: -32602, message: "Invalid params: tools/call requires a string `name`." };
    }

    try {
      return await callTool(this.client, name, params.arguments || {});
    } catch (error) {
      return {
        content: [{ type: "text", text: `Tool failed: ${error.message}` }],
        isError: true
      };
    }
  }
}
